runtime/src/master.rs

```diff
@@ -264,20 +279,46 @@ impl<T: Trait> Module<T> {
         }
 
         // execute call and collect dispatch info to return
-        let dispatch_info = proposal
+        let dispatch_result = proposal
             .clone()
-            .dispatch(system::RawOrigin::Root.into())
-            .map_err(|e| e.error)?;
+            .dispatch_bypass_filter(system::RawOrigin::Root.into());
 
         // Update round (nonce)
         Round::mutate(|round| {
             *round += 1;
         });
 
-        // events
-        Self::deposit_event(RawEvent::Executed(auth.keys().cloned().collect(), proposal));
+        // Weight from dispatch's declaration. If dispatch does not return a weight in `PostDispatchInfo`,
+        // then this weight is used.
+        let dispatch_decl_weight = proposal.get_dispatch_info().weight;
 
-        Ok(dispatch_info)
+        // Log event for success or failure of execution
+        let post_info = match dispatch_result {
+            Ok(post_dispatch_info) => {
+                Self::deposit_event(RawEvent::Executed(auth.keys().cloned().collect(), proposal));
+                post_dispatch_info
+            }
+            Err(e) => {
+                Self::deposit_event(RawEvent::ExecutionFailed(
+                    auth.keys().cloned().collect(),
+                    proposal,
+                    e.error,
+                ));
+                e.post_info
+            }
+        };
+
+        Ok(PostDispatchInfo {
+            // If weight was not given in `given_weight`, look for weight of dispatch in `post_info`. If `post_info` does not have weight,
+            // use weight from declaration. Also add minimum weight for execution
+            actual_weight: given_weight.or_else(|| {
+                Some(
+                    post_info.actual_weight.unwrap_or(dispatch_decl_weight)
+                        + get_min_weight_for_execute(&auth, T::DbWeight::get()),
+                )
+            }),
+            pays_fee: post_info.pays_fee,
+        })
```

Comment:
Alternative, maybe not superior though. https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else

```suggestion
        Ok(PostDispatchInfo {
            // If weight was not given in `given_weight`, look for weight of dispatch in `post_info`. If `post_info` does not have weight,
            // use weight from declaration. Also add minimum weight for execution
            actual_weight: Some(given_weight.unwrap_or_else(|| {
                post_info.actual_weight.unwrap_or(dispatch_decl_weight)
                    + get_min_weight_for_execute(&auth, T::DbWeight::get())
            })),
            pays_fee: post_info.pays_fee,
        })
```